python/ray/data/random_access_dataset.py

```diff
@@ -0,0 +1,262 @@
+        self._valid_blocks = []
+        self._lower_bound = None
+        self._upper_bounds = []
+        for i, b in enumerate(bounds):
+            if b:
+                self._valid_blocks.append(blocks[i])
+                if self._lower_bound is None:
+                    self._lower_bound = b[0]
+                self._upper_bounds.append(b[1])
+
+        logger.info("[setup] Creating {} random access workers.".format(num_workers))
+        self._workers = [
+            _RandomAccessWorker.options(scheduling_strategy="SPREAD").remote(
+                key, self._format
+            )
+            for _ in range(num_workers)
+        ]
+        (
+            self._block_to_workers_map,
+            self._worker_to_blocks_map,
+        ) = self._compute_block_to_worker_assignments()
+
+        logger.info(
+            "[setup] Worker to blocks assignment: {}".format(self._worker_to_blocks_map)
+        )
+        ray.get(
+            [
+                w.assign_blocks.remote(
+                    {i: self._valid_blocks[i] for i in self._worker_to_blocks_map[w]}
+                )
+                for w in self._workers
+            ]
+        )
+
+        logger.info("[setup] Finished assigning blocks to workers.")
+        self._build_time = time.perf_counter() - start
+
+    def _compute_block_to_worker_assignments(self):
+        # Return values.
+        blocks_to_workers: dict[int, List["ray.ActorHandle"]] = defaultdict(list)
+        workers_to_blocks: dict["ray.ActorHandle", List[int]] = defaultdict(list)
+
+        # Aux data structures.
+        loc_to_workers: dict[str, List["ray.ActorHandle"]] = defaultdict(list)
+        locs = ray.get([w.ping.remote() for w in self._workers])
+        for i, loc in enumerate(locs):
+            loc_to_workers[loc].append(self._workers[i])
+        block_locs = ray.experimental.get_object_locations(self._valid_blocks)
+
+        # First, try to assign all blocks to all workers at its location.
+        for block_idx, block in enumerate(self._valid_blocks):
+            block_info = block_locs[block]
+            locs = block_info.get("node_ids", [])
+            for loc in locs:
+                for worker in loc_to_workers[loc]:
+                    blocks_to_workers[block_idx].append(worker)
+                    workers_to_blocks[worker].append(block_idx)
+
+        # Randomly assign any leftover blocks to at least one worker.
+        # TODO: the load balancing here could be improved.
+        for block_idx, block in enumerate(self._valid_blocks):
+            if len(blocks_to_workers[block_idx]) == 0:
+                worker = random.choice(self._workers)
+                blocks_to_workers[block_idx].append(worker)
+                workers_to_blocks[worker].append(block_idx)
+
+        return blocks_to_workers, workers_to_blocks
+
+    def get_async(self, key: Any) -> ObjectRef[Optional[T]]:
+        """Asynchronously finds the record for a single key.
```

Comment:
What if there are multiple records with the same key?

Reply:
Clarified this in docs for to_random_access_dataset (not supported).

Reply:
I feel it'd be common to have duplicate keys, if users just pick a column and make it random-accessible.
Would it make sense to support a std::multimap<> semantic?

Reply:
Haven't seen the use case for this so far.